ERCS/erc-7786.md

```diff
@@ -0,0 +1,217 @@
+
+    function sendMessage(
+        string calldata destination, // CAIP-2 chain identifier
+        string calldata receiver, // CAIP-10 account address
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) external payable returns (bytes32 outboxId);
+}
+```
+
+#### `supportsAttribute`
+
+Returns a boolean indicating whether an attribute is supported by the gateway, identified by the selector computed from the attribute signature.
+
+A gateway MAY be upgraded with support for additional attributes. Once present support for an attribute SHOULD NOT be removed to preserve backwards compatibility with users of the gateway.
+
+#### `sendMessage`
+
+Initiates the sending of a message.
+
+Further action MAY be required by the gateway to make the sending of the message effective, such as providing payment for gas. See Post-processing.
+
+MUST revert with `UnsupportedAttribute` if an unsupported attribute key is included. MAY revert if the value of an attribute is not a valid encoding for its expected type.
+
+MAY accept call value (native token) to be sent with the message. MUST revert if call value is included but it is not a feature supported by the gateway. It is unspecified how this value is represented on the destination.
+
+MAY generate and return a unique non-zero *outbox identifier*, otherwise returning zero. This identifier shall be used to track the lifecycle of the message in the outbox in events and for post-processing.
+
+MUST emit a `MessageCreated` event, including the optional outbox identifier that is returned by the function.
+
+If an outbox identifier was generated, MUST emit a `MessageSent` event if no post-processing is required.
+
+#### `MessageCreated`
+
+This event signals that a would-be sender has initiated the sending of a message.
+
+If `outboxId` is present, post-processing MAY be required to send the message through the cross-chain channel.
+
+#### `MessageSent`
+
+This event signals that no more post-processing in the source chain is required, and that the message has been sent through the cross-chain channel.
+
+#### Post-processing
+
+After a sender has invoked `sendMessage`, further action MAY be required by the gateways to make the message effective. This is called *post-processing*. For example, some payment is typically required to cover the gas of executing the message at the destination.
+
+The exact interface for any such action is out of scope of this ERC. If the `postProcessingOwner` attribute is supported and present, such actions MUST be restricted to the specified account, otherwise they MUST be able to be performed by any party in a way that MUST NOT be able to compromise the eventual receipt of the message.
+
+The gateway MUST emit a `MessageSent` event with the appropriate identifier once no more post-processing is required by the source gateway. Further post-processing MAY be required in the source or destination gateways.
```

Comment:
I think we need to distinguish protocol post-processing vs other post processing. In the Axelar example, gas payment is required and this is a form of post processing, but the gateway has no knowledge of this. So it's unclear based on how this is written if/when the gateway should emit `MessageSent`.

I personally think `MessageSent` could probably use a renaming, and simply means the message is ready for sending. Whether the message is actually sent or not is hard for the gateway to determine and might not be possible, but the gateway does know if anything else needs to be done for the message to be able to be sent.

Reply:
What do you think about `MessageReady`. What other terms would you propose?